cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_hist_impl_dpc.hpp

```diff
@@ -0,0 +1,2966 @@
+        imp -= Float(cls_count) * (cls_count)*div;
+
+        if (cls_count > max_cls_count) {
+            max_cls_count = cls_count;
+            win_cls = cls_idx;
+        }
+    }
+
+    node_ptr[5] = win_cls;
+    node_imp_ptr[0] = sycl::max(imp, Float(0));
+}
+
+// regression compute_hist_for_node
+template <typename Float, typename Index>
+inline void compute_hist_for_node(
+    sycl::nd_item<2> item,
+    Index ind_start,
+    Index ind_end,
+    const Float* response_ptr,
+    Index* node_ptr,
+    const Index* node_tree_order_ptr,
+    typename task_types<Float, Index, task::regression>::hist_type_t* local_buf_ptr,
+    const imp_data_list_ptr_mutable<Float, Index, task::regression>& imp_list_ptr,
+    const kernel_context<Float, Index, task::regression>& ctx,
+    Index node_id) {
+    using task_t = task::regression;
+    using impl_const_t = impl_const<Index, task_t>;
+    using hist_type_t = typename task_types<Float, Index, task_t>::hist_type_t;
+
+    const Index local_id = item.get_local_id()[0];
+    const Index local_size = item.get_local_range()[0];
+
+    Float* node_imp_ptr = imp_list_ptr.imp_list_ptr_ + node_id * impl_const_t::node_imp_prop_count_;
+    constexpr Index hist_prop_count = impl_const_t::hist_prop_count_;
+    hist_type_t private_histogram[hist_prop_count] = { 0 };
+
+    for (Index i = ind_start; i < ind_end; ++i) {
+        Index id = node_tree_order_ptr[i];
+        add_val_to_hist<Float, Index>(private_histogram, response_ptr[id]);
+    }
+
+    hist_type_t* local_h_ptr = local_buf_ptr + local_id * hist_prop_count;
+
+    local_h_ptr[0] = private_histogram[0];
+    local_h_ptr[1] = private_histogram[1];
+    local_h_ptr[2] = private_histogram[2];
+
+    for (Index offset = local_size / 2; offset > 0; offset >>= 1) {
+        item.barrier(sycl::access::fence_space::local_space);
+        if (local_id < offset) {
+            hist_type_t* h_ptr = local_buf_ptr + (local_id + offset) * hist_prop_count;
+            merge_stat(local_h_ptr, h_ptr, hist_prop_count);
+        }
+    }
+
+    if (local_id == 0) {
+        node_imp_ptr[0] = local_h_ptr[1]; // store mean
+        node_imp_ptr[1] = local_h_ptr[2]; // store sum2cent
+    }
+}
+
+template <typename Float, typename Bin, typename Index, typename Task>
+sycl::event train_kernel_hist_impl<Float, Bin, Index, Task>::compute_initial_imp_for_node_list(
+    const train_context_t& ctx,
+    imp_data_t& imp_data_list,
+    pr::ndarray<Index, 1>& node_list,
+    Index node_count,
+    const bk::event_vector& deps) {
+    ONEDAL_ASSERT(imp_data_list.imp_list_.get_count() ==
+                  node_count * impl_const_t::node_imp_prop_count_);
+    if constexpr (std::is_same_v<task::classification, Task>) {
+        ONEDAL_ASSERT(imp_data_list.class_hist_list_.get_count() == node_count * ctx.class_count_);
+    }
+    ONEDAL_ASSERT(node_list.get_count() == node_count * impl_const_t::node_prop_count_);
+
+    sycl::event last_event;
```

Comment:
Do you need a variable to return empty event?

Reply:
removed